core/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceService.java

```diff
@@ -113,7 +113,22 @@ private RecoveryResponse recover(final StartRecoveryRequest request) throws IOEx
             throw new DelayRecoveryException("source node has the state of the target shard to be [" + targetShardRouting.state() + "], expecting to be [initializing]");
         }
 
-        RecoverySourceHandler handler = ongoingRecoveries.addNewRecovery(request, targetShardRouting.allocationId().getId(), shard);
+        // the following IF condition won't go into master, it only goes into 5.4 / 5.x branch to allow for interoperability with < 5.4
```

Comment:
At which point will you also put back the `final` modifier on the `request` method parameter?